src/stages/normalize/patchers/FunctionPatcher.js

```diff
@@ -0,0 +1,57 @@
+import NodePatcher from './../../../patchers/NodePatcher.js';
+import type { Node, ParseContext, Editor } from './../../../patchers/types.js';
+
+export default class FunctionPatcher extends NodePatcher {
+  parameters: Array<NodePatcher>;
+  body: ?NodePatcher;
+  
+  constructor(node: Node, context: ParseContext, editor: Editor, parameters: Array<NodePatcher>, body: ?NodePatcher) {
+    super(node, context, editor);
+    this.parameters = parameters;
+    this.body = body;
+  }
+
+  patchAsExpression() {
+    // To avoid knowledge of all the details how assignments can be nested in nodes,
+    // we add a callback to the function node before patching the parameters and remove it afterwards.
+    // This is detected and used by the MemberAccessOpPatcher to claim a free binding for this parameter
+    // (from the functions scope, not the body's scope)
+
+    let assignments = [];
+    this.addStatementAtScopeHeader = (memberName: string) => {
+      let varName = this.claimFreeBinding(memberName);
+      assignments.push(`@${memberName} = ${varName}`);
+      this.log(`Replacing parameter @${memberName} with ${varName}`);
+      return varName;
+    };
+
+    this.parameters.forEach(parameter => parameter.patch());
+
+    delete this.addStatementAtScopeHeader;
+
+    // If there were assignments from parameters insert them
+    if (this.body) {
+      // before the actual body
+      if (assignments.length) {
+        let text;
+        if (this.body.node.inline) {
```

Comment:
Which lives in the main stage… I ran into trouble trying all this in the main stage with the editor complaining about editing positions twice, if I remember right. Maybe this was, because the actual parameter (e.g. `@a`) could be processed in different ways from other patchers…